frontend/pkg/frontend/frontend.go

```diff
@@ -431,6 +509,227 @@ func (f *Frontend) ArmResourceCreateOrUpdate(writer http.ResponseWriter, request
 	}
 }
 
+// clusterResourceIdFromNodePoolResourceID strips the nodepool part of the full resource ID
+// returning the full cluster resourceID.
+// given a full resource ID like
+// /subscriptions/00000000-0000-0000-0000-000000000000/resourcegroups/dev-test-rg/providers/microsoft.redhatopenshift/hcpopenshiftclusters/aldo-cluster/nodepools/aldo-nodepool
+// it returns the resourceID of the cluster
+// /subscriptions/00000000-0000-0000-0000-000000000000/resourcegroups/dev-test-rg/providers/microsoft.redhatopenshift/hcpopenshiftclusters/aldo-cluster
+func clusterResourceIdFromNodePoolResourceID(resourceID string) string {
+	i := strings.Index(resourceID, "/nodepools")
+	clusterResourceId := resourceID[0:i]
+	return clusterResourceId
+}
+
+func (f *Frontend) CreateNodePool(writer http.ResponseWriter, request *http.Request) {
+	var err error
+
+	ctx := request.Context()
+
+	versionedInterface, err := VersionFromContext(ctx)
+	if err != nil {
+		f.logger.Error(err.Error())
+		arm.WriteInternalServerError(writer)
+		return
+	}
+
+	systemData, err := SystemDataFromContext(ctx)
+	if err != nil {
+		f.logger.Error(err.Error())
+		arm.WriteInternalServerError(writer)
+		return
+	}
+
+	f.logger.Info(fmt.Sprintf("%s: CreateNodePool", versionedInterface))
+
+	// URL path is already lowercased by middleware.
+	nodePoolResourceID := request.URL.Path
+
+	clusterResourceID := clusterResourceIdFromNodePoolResourceID(nodePoolResourceID)
+
+	subscriptionID := request.PathValue(PathSegmentSubscriptionID)
+	clusterDoc, err := f.dbClient.GetClusterDoc(ctx, clusterResourceID, subscriptionID)
+	if err != nil {
+		if errors.Is(err, database.ErrNotFound) {
+			f.logger.Info(fmt.Sprintf("existing document not found for cluster %s when creating node pool", clusterResourceID))
+			arm.WriteInternalServerError(writer)
+			return
+		}
+		f.logger.Error(fmt.Sprintf("failed to fetch cluster document for %s when creating node pool: %v", clusterResourceID, err))
+		arm.WriteInternalServerError(writer)
+		return
+	}
+
+	nodePoolDoc, err := f.dbClient.GetNodePoolDoc(ctx, nodePoolResourceID)
+	if err != nil {
+		if errors.Is(err, database.ErrNotFound) {
+			f.logger.Info(fmt.Sprintf("creating nodepool document for %s", nodePoolResourceID))
+
+			nodePoolDoc = &database.NodePoolDocument{
+				ID:           uuid.New().String(),
+				Key:          nodePoolResourceID,
+				PartitionKey: subscriptionID,
+				SystemData:   systemData,
+			}
+		} else {
+			f.logger.Error(fmt.Sprintf("failed to fetch node pool document for %s: %v", nodePoolResourceID, err))
+			arm.WriteInternalServerError(writer)
+			return
+		}
+	}
+
+	if clusterDoc.ClusterID == "" {
+		f.logger.Error("unexpected error: clusterID of clusterDoc is empty.")
```

Comment:
can we log out clusterID?

Reply:
Sure. That was the idea. Good catch 🙂 Thanks!

Reply:
ack this is an essentially: "should never happen" sort of thing.  My bad.